### PR TITLE
Extending signalr-go-server for async hub methods and methods with multiple results

### PR DESCRIPTION
**Pull request by philippseith**
- Invocation of hub methods which return a chan T. It is waited for the channel to return a single result in a separate goroutine , then the completion is sent
- Simple extension for methods returning multiple results

### CODE REVIEW

File: signalr-go-server/cmd/main.go [hunk 1]
```diff
@@ -28,13 +29,26 @@ func (c *chat) Send(message string) {
...
+
 func main() {
 	router := http.NewServeMux()
 	router.Handle("/", http.FileServer(http.Dir("../public")))
 
 	signalr.MapHub(router, "/chat", &chat{})
 
-	if err := http.ListenAndServe("localhost:8086", router); err != nil {
+	if err := http.ListenAndServe("localhost:8087", router); err != nil {
 		log.Fatal("ListenAndServe:", err)
```
Comment (philippseith) [inline signalr-go-server/cmd/main.go]: Ignore that :-)

File: signalr-go-server/pkg/signalr/signalr.go [hunk 2]
```diff
@@ -518,6 +523,20 @@ func hubConnectionHandler(connectionID string, ws *websocket.Conn, hubInfo *hubI
 	waitgroup.Wait()
 }
 
+func doCompletion(conn webSocketHubConnection, invocation hubInvocationMessage, result []reflect.Value) {
+	if len(result) > 0 {
+		// REVIEW: When is this ever > 1
+		values := make([]interface{}, 0, len(result))
```
Comment (philippseith) [inline signalr-go-server/pkg/signalr/signalr.go]: For methods returning tuples, it is

File: signalr-go-server/pkg/signalr/signalr.go [hunk 3]
```diff
@@ -518,6 +523,20 @@ func hubConnectionHandler(connectionID string, ws *websocket.Conn, hubInfo *hubI
 	waitgroup.Wait()
 }
 
+func doCompletion(conn webSocketHubConnection, invocation hubInvocationMessage, result []reflect.Value) {
+	if len(result) > 0 {
+		// REVIEW: When is this ever > 1
```
Comment (philippseith) [inline signalr-go-server/pkg/signalr/signalr.go]: len(result) is the tuple size

File: signalr-go-server/pkg/signalr/hubclients.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
+package signalr
+
+type HubClients interface {
+	All() ClientProxy
+	Client(connectionID string) ClientProxy
+	Group(groupName string) ClientProxy
+	Caller() ClientProxy
```
Comment (davidfowl) [inline signalr-go-server/pkg/signalr/hubclients.go]: We don't do caller unfortunately unless we have a per invocation way of passing this state. This is still a gap but we should remove it for now.

File: signalr-go-server/pkg/signalr/hublifetimemanager.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+type HubLifetimeManager interface {
+	OnConnected(conn hubConnection)
+	OnDisconnected(conn hubConnection)
+	InvokeAll(target string, args []interface{})
+	InvokeClient(connectionID string, target string, args []interface{})
+	InvokeGroup(groupName string, target string, args []interface{})
+	AddToGroup(groupName, connectionID string)
+	RemoveFromGroup(groupName, connectionID string)
+}
+type hubConnection interface {
```
Comment (davidfowl) [inline signalr-go-server/pkg/signalr/hublifetimemanager.go]: nit: space

File: signalr-go-server/pkg/signalr/hubprotocol.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+import (
+	"bytes"
+	"io"
+)
+
+// HubProtocol interface
+type HubProtocol interface {
+	ReadMessage(buf *bytes.Buffer) (interface{}, error)
+	WriteMessage(message interface{}, writer io.Writer) error
+	UnmarshalArgument(argument interface{}, value interface{}) error
```
Comment (davidfowl) [inline signalr-go-server/pkg/signalr/hubprotocol.go]: This is a bit different in the other clients/servers and instead is another interface. Lets leave this for now to though.

File: signalr-go-server/pkg/signalr/server.go [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+	mux.Handle(path, websocket.Handler(func(ws *websocket.Conn) {
+		connectionID := ws.Request().URL.Query().Get("id")
+
+		if len(connectionID) == 0 {
+			// Support websocket connection without negotiate
+			connectionID = getConnectionID()
+		}
+
+		// Copy hubPrototype
+		hub := reflect.New(reflect.ValueOf(hubPrototype).Elem().Type()).Interface().(HubInterface)
```
Comment (davidfowl) [inline signalr-go-server/pkg/signalr/server.go]: Why?

File: signalr-go-server/pkg/signalr/streamclient.go [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+				upChan.Send(reflect.ValueOf(fmt.Sprint(f)))
+			}
+		} else {
+			upChan.Send(reflect.ValueOf(streamItem.Item))
+		}
+	}
+}
+
+func (u *streamClient) receiveCompletionItem(completion completionMessage) {
+	channel := u.upstreamChannels[completion.InvocationID]
```
Comment (davidfowl) [inline signalr-go-server/pkg/signalr/streamclient.go]: We need to handle the case where this invocation id is invalid.

File: signalr-go-server/pkg/signalr/server.go [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+					// Unable to find the method
+					conn.completion(invocation.InvocationID, nil, fmt.Sprintf("Unknown method %s", invocation.Target))
+				} else if in, clientStreaming, err := buildMethodArguments(method, invocation, streamClient, protocol); err != nil {
+					// argument build failed
+					conn.completion(invocation.InvocationID, nil, err.Error())
+				} else if clientStreaming {
+					// let the receiving method run idependently
+					go func() {
+						defer func() {
+							if err := recover(); err != nil {
```
Comment (davidfowl) [inline signalr-go-server/pkg/signalr/server.go]: What's recover()?

**Comment by davidfowl** [comment]
@philippseith are you planning to make those suggested changes?

**Comment by davidfowl** [comment]
@philippseith Are you planning to contribute this back? I'd love to merge the PR but it seems as though you're trying to make a standalone repo for the go server impl (which I'd be open to as well).

File: signalr-go-server/cmd/main.go (hunk 1 above)
Comment (davidfowl) [reply signalr-go-server/cmd/main.go]: Are you intending to push this change upstream?

File: signalr-go-server/pkg/signalr/server.go (hunk 7 above)
Comment (philippseith) [reply signalr-go-server/pkg/signalr/server.go]: To implement ClientProxy.Caller(). Each connection get an hub instance

File: signalr-go-server/cmd/main.go (hunk 1 above)
Comment (philippseith) [reply signalr-go-server/cmd/main.go]: No

File: signalr-go-server/pkg/signalr/hubclients.go (hunk 4 above)
Comment (philippseith) [reply signalr-go-server/pkg/signalr/hubclients.go]: see code for copying the hub prototype below

File: signalr-go-server/pkg/signalr/server.go (hunk 9 above)
Comment (philippseith) [reply signalr-go-server/pkg/signalr/server.go]: deferred functions are called after the surrounding block ends, even if the block panicked. recover() then returns the panics error and stack

File: signalr-go-server/pkg/signalr/server.go (hunk 7 above)
Comment (davidfowl) [reply signalr-go-server/pkg/signalr/server.go]: Yea, I'm jaded about this. If creation of the hub happens per invocation then it requires dependency injection to pass state into the hub. I don't think we should implement caller as yet. That was explicitly left out.

**Comment by philippseith** [comment]
My final target is to make a standalone repo, because it will be more useful for other go projects. But i might create a new branch with the code changes but without the moving and renaming and open a new pull request  for it.

**Comment by davidfowl** [comment]
> My final target is to make a standalone repo, because it will be more useful for other go projects. But i might create a new branch with the code changes but without the moving and renaming and open a new pull request for it.

I'd like to own that repo so lets coordinate here. Can we first work on getting your contribution back into this repo then work on making a repo that can be used as a golang package?

**Comment by philippseith** [comment]
> I'd like to own that repo so lets coordinate here. Can we first work on getting your contribution back into this repo then work on making a repo that can be used as a golang package?

Yes, we can. I will close this pull request and open another where structure is preserved

**Pull request closed without merging**